strapi/strapi
# Api token v2/permissions for api token

### What does it do?

- adds `custom` token type
- adds `permissions` array of actions to custom tokens
- adds `lastUsed` timestamp to api tokens
- api tokens authorization now checks custom token permissions
- api tokens authorization now updates `lastUsed` time when authorized

There are still a few areas with uncertainty, but I think it's worth merging in and fixing them later if necessary. Without this we're blocked in a few areas, since the regenerate feature is based off this branch to prevent conflicts. We can also integrate the front-end and QA can start e2e testing.

### Why is it needed?

To make API Tokens v2 work with custom permissions

### How to test it?

- Create/update/delete custom tokens using api-token service OR the admin api
- check if they work with the content api

```js
    // Create a token
    const payload = {
      type: 'custom',
      permissions: ['api::restaurant.restaurant.find'],
      name: 'custom token v2',
    };
    const token = await strapi.admin.services['api-token'].create(payload);

    // update a token
    const updatedToken = await strapi.admin.services['api-token'].update(token.id, {
      description: 'foobar',
      permissions: ['api::restaurant.restaurant.find', 'api::restaurant.restaurant.findOne'],
    });
```

Tests have been provided and existing tests updated:

```sh
# runs the authorization, controller, and service unit tests:
yarn test:unit api-token
# runs the e2e tests
node ./test/e2e.js api-token
```

### Related issue(s)/PR(s)

Front end for last used timestamp:
https://github.com/strapi/strapi/pull/14108
